<div style="background:linear-gradient(135deg,#1e1b4b 0%,#4338ca 55%,#6366f1 100%);border-radius:18px;padding:34px 30px;color:#fff;font-family:Inter,Segoe UI,sans-serif">
  <div style="font-size:12px;letter-spacing:3px;color:#c7d2fe;font-weight:700;text-transform:uppercase">Chapter 34 · Solutions</div>
  <div style="font-size:34px;font-weight:900;line-height:1.1;margin:10px 0 6px">Bayes' Theorem, Worked Answers ✅</div>
  <div style="font-size:15px;color:#e0e7ff;max-width:720px;line-height:1.6">Full solutions to the five challenges. Try them yourself first, then compare.</div>
  <div style="margin-top:16px;font-size:13px;color:#cbd5e1">Statistics, Data Science and AI: A Visual Handbook · John Fisher · 2026</div>
</div>

## ⚙️ Setup

In [1]:
import numpy as np
rng = np.random.default_rng(341)
print("ready")

ready


<div style="background:#eef2ff;border-left:5px solid #4f46e5;border-radius:10px;padding:16px 20px;font-family:Inter,sans-serif">
<span style="font-size:13px;font-weight:800;color:#4f46e5;letter-spacing:1px">CHALLENGE 1 · REVERSE A CONDITIONAL</span>
<div style="font-size:21px;font-weight:800;color:#1a2138;margin-top:4px">Bayes from a tree</div>
<div style="color:#4a5578;margin-top:6px">70% of email is legitimate, 30% is spam. The word 'offer' appears in 60% of spam and 5% of legit mail. An email contains 'offer'. Find P(spam | 'offer').</div>
</div>

In [2]:
p_spam, p_legit = 0.30, 0.70
p_off_spam, p_off_legit = 0.60, 0.05
p_off = p_spam*p_off_spam + p_legit*p_off_legit     # total probability
post  = p_spam*p_off_spam / p_off                    # Bayes
print(f"P(\"offer\")           = {p_off:.3f}")
print(f"P(spam | \"offer\")    = {post:.3f}")

P("offer")           = 0.215
P(spam | "offer")    = 0.837


**Answer:** P("offer") = 0.30(0.60) + 0.70(0.05) = 0.18 + 0.035 = 0.215. Then P(spam | "offer") = 0.18 / 0.215 &approx; **0.837**. One suggestive word lifts the spam probability from a 0.30 prior to about 84%.

<div style="background:#eef2ff;border-left:5px solid #4f46e5;border-radius:10px;padding:16px 20px;font-family:Inter,sans-serif">
<span style="font-size:13px;font-weight:800;color:#2563eb;letter-spacing:1px">CHALLENGE 2 · LAW OF TOTAL PROBABILITY</span>
<div style="font-size:21px;font-weight:800;color:#1a2138;margin-top:4px">Average over three suppliers</div>
<div style="color:#4a5578;margin-top:6px">Three suppliers send parts: S1 = 50% of stock at 1% defective, S2 = 30% at 2%, S3 = 20% at 3%. What fraction of all parts are defective?</div>
</div>

In [3]:
w   = np.array([0.50, 0.30, 0.20])     # share from each supplier
d   = np.array([0.01, 0.02, 0.03])     # defect rate of each
p_def = (w*d).sum()
for i,(wi,di) in enumerate(zip(w,d),1):
    print(f"S{i}: {wi:.2f} * {di:.2f} = {wi*di:.3f}")
print(f"P(defective) = {p_def:.3f}")

S1: 0.50 * 0.01 = 0.005
S2: 0.30 * 0.02 = 0.006
S3: 0.20 * 0.03 = 0.006
P(defective) = 0.017


**Answer:** the Law of Total Probability averages the defect rates, weighted by each supplier's share: P(defective) = 0.50(0.01) + 0.30(0.02) + 0.20(0.03) = 0.005 + 0.006 + 0.006 = **0.017**, that is 1.7%. The suppliers partition every part, so the slices simply add.

<div style="background:#eef2ff;border-left:5px solid #4f46e5;border-radius:10px;padding:16px 20px;font-family:Inter,sans-serif">
<span style="font-size:13px;font-weight:800;color:#059669;letter-spacing:1px">CHALLENGE 3 · PRIOR &rarr; POSTERIOR</span>
<div style="font-size:21px;font-weight:800;color:#1a2138;margin-top:4px">A single test</div>
<div style="color:#4a5578;margin-top:6px">A condition has 2% prevalence. A test has 90% sensitivity and 85% specificity. Given one positive result, find P(condition | positive).</div>
</div>

In [4]:
prior, sens, spec = 0.02, 0.90, 0.85
p_pos = prior*sens + (1-prior)*(1-spec)
post  = prior*sens / p_pos
print(f"P(positive)             = {p_pos:.3f}")
print(f"P(condition | positive) = {post:.3f}  (about {post:.0%})")

P(positive)             = 0.165
P(condition | positive) = 0.109  (about 11%)


**Answer:** P(positive) = 0.02(0.90) + 0.98(0.15) = 0.018 + 0.147 = 0.165. Posterior = 0.018 / 0.165 &approx; **0.109**, about 11%. Even a positive test leaves the condition unlikely, because a 15% false-positive rate on the large healthy group produces many more false alarms than true ones.

<div style="background:#eef2ff;border-left:5px solid #4f46e5;border-radius:10px;padding:16px 20px;font-family:Inter,sans-serif">
<span style="font-size:13px;font-weight:800;color:#db2777;letter-spacing:1px">CHALLENGE 4 · SEQUENTIAL UPDATING</span>
<div style="font-size:21px;font-weight:800;color:#1a2138;margin-top:4px">A second positive test</div>
<div style="color:#4a5578;margin-top:6px">The patient from Challenge 3 takes a second, independent test that is also positive. Update again, using the previous posterior as the new prior.</div>
</div>

In [5]:
prior = 0.109                 # the posterior from Challenge 3
sens, spec = 0.90, 0.85
p_pos = prior*sens + (1-prior)*(1-spec)
post  = prior*sens / p_pos
print(f"new prior (was posterior) = {prior:.3f}")
print(f"P(condition | 2 positives) = {post:.3f}  (about {post:.0%})")

new prior (was posterior) = 0.109
P(condition | 2 positives) = 0.423  (about 42%)


**Answer:** with the new prior 0.109: P(positive) = 0.109(0.90) + 0.891(0.15) = 0.0981 + 0.1337 = 0.2318, so the posterior = 0.0981 / 0.2318 &approx; **0.423**, about 42%. Two positives are far more convincing than one: belief climbed from 2% to 11% to 42% as independent evidence accumulated.

<div style="background:#eef2ff;border-left:5px solid #4f46e5;border-radius:10px;padding:16px 20px;font-family:Inter,sans-serif">
<span style="font-size:13px;font-weight:800;color:#7c3aed;letter-spacing:1px">CHALLENGE 5 · NAIVE BAYES</span>
<div style="font-size:21px;font-weight:800;color:#1a2138;margin-top:4px">P(class | feature)</div>
<div style="color:#4a5578;margin-top:6px">A spam filter has prior P(spam) = 0.5. The word 'win' appears in 70% of spam and 5% of ham. Given 'win', find P(spam | 'win').</div>
</div>

In [6]:
p_spam = 0.50
p_win_spam, p_win_ham = 0.70, 0.05
p_win = p_spam*p_win_spam + (1-p_spam)*p_win_ham
post  = p_spam*p_win_spam / p_win
print(f"P(\"win\")          = {p_win:.3f}")
print(f"P(spam | \"win\")   = {post:.3f}  (about {post:.0%})")

P("win")          = 0.375
P(spam | "win")   = 0.933  (about 93%)


**Answer:** P("win") = 0.5(0.70) + 0.5(0.05) = 0.35 + 0.025 = 0.375. Posterior = 0.35 / 0.375 &approx; **0.933**, about 93%. The strong likelihood ratio (0.70 vs 0.05) makes "win" a powerful spam signal; a full classifier multiplies many such word likelihoods together.

---
<div style="background:#ffffff;border:1px solid #e6e9f2;border-radius:16px;padding:22px 26px;font-family:Inter,sans-serif">
<div style="font-size:19px;font-weight:800;color:#1a2138">🎉 Nicely done</div>
<div style="color:#4a5578;line-height:1.8;margin-top:8px">You reversed conditionals with Bayes, averaged over a partition with the Law of Total Probability, updated a prior into a posterior, chained two updates as evidence accumulated, and scored a class with naive Bayes. The pattern never changed: posterior &prop; likelihood &times; prior.</div>
</div>

<div style="text-align:center;color:#8b94b3;font-size:12px;margin-top:16px">Statistics, Data Science and AI: A Visual Handbook · © 2026 John Fisher</div>